python_modules/libraries/dagster-dg-cli/dagster_dg_cli/cli/scaffold/branch/diagnostics.py

```diff
@@ -0,0 +1,315 @@
+
+    def __init__(
+        self,
+        level: DiagnosticsLevel = "off",
+        output_dir: Optional[Path] = None,
+        correlation_id: Optional[str] = None,
+    ):
+        self.level = level
+        self.correlation_id = correlation_id or str(uuid.uuid4())
+        self.output_dir = output_dir or Path(tempfile.gettempdir()) / "dg" / "diagnostics"
+        self.entries: list[DiagnosticsEntry] = []
+        self._output_file: Optional[Path] = None
+
+        # Ensure output directory exists and create output file if diagnostics are enabled
+        if self.level != "off":
+            self.output_dir.mkdir(parents=True, exist_ok=True)
+            timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
+            filename = f"scaffold_diagnostics_{timestamp}_{self.correlation_id[:8]}.json"
+            self._output_file = self.output_dir / filename
+
+            # Initialize file with session metadata
+            session_metadata = {
+                "correlation_id": self.correlation_id,
+                "session_start": datetime.now().isoformat(),
+                "level": self.level,
+                "entries": [],
+            }
+            self._output_file.write_text(json.dumps(session_metadata, indent=2))
+
+    @property
+    def output_file(self) -> Optional[Path]:
+        """Get the current output file path."""
+        return self._output_file
+
+    def _should_log(self, entry_level: DiagnosticsLevel) -> bool:
+        """Check if entry should be logged based on current diagnostics level."""
+        if self.level == "off":
+            return False
+
+        # Define hierarchy as mapping for cleaner logic
+        level_priority = {"error": 0, "info": 1, "debug": 2}
+
+        current_priority = level_priority.get(self.level)
+        entry_priority = level_priority.get(entry_level)
+
+        # Both levels must be valid for logging to proceed
+        if current_priority is None or entry_priority is None:
+            return False
+
+        return entry_priority <= current_priority
+
+    def log(
+        self,
+        level: DiagnosticsLevel,
+        category: str,
+        message: str,
+        data: Optional[dict[str, Any]] = None,
+    ) -> None:
+        """Log a diagnostics entry."""
+        if not self._should_log(level):
+            return
+
+        entry = DiagnosticsEntry(
+            correlation_id=self.correlation_id,
+            timestamp=datetime.now().isoformat(),
+            level=level,
+            category=category,
+            message=message,
+            data=data or {},
+        )
+
+        self.entries.append(entry)
+
+        # Stream entry to file immediately if file exists
+        if self._output_file and self._output_file.exists():
+            self._append_entry_to_file(entry)
+
+    def _append_entry_to_file(self, entry: DiagnosticsEntry) -> None:
+        """Append a single entry to the JSON file by rewriting it."""
```

Comment:
Not sure how many entries we're expecting to generate or how fast they will come but seems less than ideal to regen the whole file for every entry

Reply:
thanks good catch

Reply:
will change to streaming json